File: Java/SerialMapReduce/src/ClassificationTask.java
import java.math.BigDecimal;
import java.math.BigInteger;
import java.math.MathContext;
import java.util.Arrays;
import java.util.HashMap;
import java.util.concurrent.atomic.AtomicInteger;

import org.ietf.jgss.Oid;

class Classification{
	static int trainingSetCount;
	int id;
	String label; // this classifciation belongs to label
	BigDecimal featureMultiplication = BigDecimal.ONE; // Multiplication of this features
	int featureCount = 0; // How many features we counted
	final int labelCount; // How many times this label repeats in the set
	boolean finalized = false;
	double res;
	
	public Classification(int id, String label, int labelCount) {
		this.id = id;
		this.label = label;
		this.labelCount = labelCount;
	}

	public void addFeature(int valFeatureModel) {
		featureCount++;
		featureMultiplication = featureMultiplication.multiply(new BigDecimal((valFeatureModel + .007) + ""));
		
	}
	
	public double end(){
		if (finalized)
			return res;
		if (featureCount < 2)
			return 0;
		BigDecimal featurePower = new BigDecimal(labelCount + "");
		featurePower.pow(featureCount - 1);
		double resL =  featureMultiplication.divide(featurePower.multiply(new BigDecimal(trainingSetCount + "")), MathContext.DECIMAL128).doubleValue();
		res = resL;
		finalized = true;
		return resL;
	}
	
	@Override
	public String toString() {
		String probs = "Tuple #" + id + " prob of being " + label + " is " + end() + "\n";
		return probs;
	}
	
}

// READ-ONLY class that represent a map task
public class ClassificationTask{
	int id = 0;
	String[] attrs = null;
	String originalLine = null;
	boolean initialized = false;
	HashMap<String, Classification> classification = new HashMap<String, Classification>();
	
	public ClassificationTask(String line, int id) {
		this.id = id;
		this.originalLine = line;
	}

	public void initializeMe(){
		if (!initialized){
			attrs = originalLine.split("\t");
			initialized = true;
			for(String label : Globals.categories){
				int valLabelOverall = Globals.modelDict.get(label + "_class").intValue(); // Count of this label overall
				classification.put(label, new Classification(id, label, valLabelOverall));
				Classification.trainingSetCount += valLabelOverall;
			}
		}
	}

	@Override
	public String toString() {
		StringBuilder str = new StringBuilder();
		String winner = "";
		double winnerVal = Double.NEGATIVE_INFINITY;
		for (String label : classification.keySet()){
			str.append(classification.get(label).toString());
			if (classification.get(label).end() > winnerVal){
				winner = label;
				winnerVal = classification.get(label).end();
			}
		}
		String fin = "So, the biggest is: " + winner + " : " + winnerVal + "\n";
		return str.toString() + fin;
	}
	
	public String[] getAttrs() {
		initializeMe();
		return attrs;
	}

	public void classificate() {
		initializeMe();
		String[] attr = this.getAttrs();
		features: for(int i = 0; i < 600; i++){
			String featureValue = attr[i];
			if (featureValue.charAt(0) == '1'){
				for(String label : Globals.categories){
					Classification classTmp = classification.get(label);
					
					AtomicInteger valFeatureModel = Globals.modelDict.get(label + "_" + i); // Count of this feature in this label
					if (valFeatureModel == null) // This feature was unknown always, skip it
						classTmp.addFeature(0);
					else
						classTmp.addFeature(valFeatureModel.intValue());
					
				}
			}
		}
	}

}

File: Cplusplus/Parallel/Classification_task.cpp
#include "Classification_task.h"

int Classification::trainingSetCount = 0;

Classification::Classification(int id, string label, int labelCount){
	this->id = id;
	this->label = label;
	this->labelCount = labelCount;
}

void Classification::addFeature(int valFeatureModel){
	featureCount++;
	featureMultiplication *= valFeatureModel + .007;
}

double Classification::end(){
	if (finalized)
		return res;
	if (featureCount < 2)
		return 0;
	unsigned long long featurePower = labelCount;
	featurePower = pow(featurePower, featureCount - 1);

	double resL = featureMultiplication / (featurePower * (trainingSetCount));
	if (!_finite(resL)){
		if (resL > 0) 
			return numeric_limits<double>::max();
		return numeric_limits<double>::max() * -1;
	}

	res = resL;
	finalized = true;
	return resL;
}

string Classification::toString(){
	string probs("Tuple #" + to_string(id) + " prob of being " + label + " is " + to_string(end()) + "\n");
	return probs;
}

///////////////////////////////////////////////////////////////


Classification_task::Classification_task(string& line, int i)
{
	originalLine.assign(line);
	id = i;
	attrs = vector<int>();
	classes = Classifications();
	initialized = false;
}

void Classification_task::initialize()
{
	int i = id;
	if (!initialized){
		split(originalLine, '\t');
		for (int i = 0; i < MAX_CATEGORIES; i++){
			Model_ReduceHashMap::const_accessor a;
			string label(categories[i]);
			modelDictionary->find(a, label + "_class");
			int valLabelOverall = a->second;
			Classification tmp(id, label, valLabelOverall);
			pair<string, Classification> p(label, tmp);
			classes.insert(p);
			Classification::trainingSetCount += valLabelOverall;
		}
	}
}

void Classification_task::classificate()
{
	initialize();
	for (int i = 0; i < 50; i++){
		int featureValue = attrs[i];
		if (featureValue == 1){
			for (int i = 0; i < MAX_CATEGORIES; i++){
				string label(categories[i]);
				Classifications::iterator got = classes.find(label);
				Classification classTmp = got->second;

				Model_ReduceHashMap::const_accessor a;
				string find = label + "_" + to_string(i);
				if (!modelDictionary->find(a, find)){ // This feature was unknown always, skip it
					classTmp.addFeature(0);
				}
				else{
					int valFeatureModel = a->second; // Count of this feature in this label
					classTmp.addFeature(valFeatureModel);
				}
				got->second = classTmp;

			}
		}
	}
}

string Classification_task::toString(){
	string str;
	string winner;
	double winnerVal = -1 * numeric_limits<double>::max();
	for (Classifications::iterator it = classes.begin();
		it != classes.end();
		it++){
		str.append(it->second.toString());
		if (it->second.end() > winnerVal){
			winner = it->first;
			winnerVal = it->second.end();
		}
	}
	string fin = "So, the biggest is: " + winner + " : " + to_string(winnerVal) + "\n";
	return str.append(fin);
}



// Created by: http://stackoverflow.com/users/7980/alec-thomas
// Adapted by: http://github.com/kennyeni
void Classification_task::split(const string &text, const char sep){
	//int start = 0, end = 0;
	char tmp;
	int i = 0;
	int offset = 0;
	const char * line = text.c_str();
	while (sscanf(line + offset, "%c%n", &tmp, &i) == 1){
		if (tmp == '1' || tmp == '0' || tmp == '?'){
			attrs.push_back(tmp - '0');
		}
		offset += i;
	}
}